daemon/src/lib.rs

```diff
@@ -83,6 +87,89 @@ pub const PING_INTERVAL: Duration = Duration::from_secs(30);
 
 pub const N_PAYOUTS: usize = 200;
 
+#[derive(Clone, Copy)]
+pub struct ProtocolFactory;
+
+impl ProtocolFactory {
+    const PONG_V1: &'static str = xtra_libp2p_ping::PROTOCOL_NAME;
+    const IDENTIFY_V1: &'static str = identify::PROTOCOL;
+    const OFFERS_V1: &'static str = xtra_libp2p_offer::PROTOCOL_NAME;
+    const ROLLOVER_V1: &'static str = rollover::PROTOCOL;
+    const COLLAB_SETTLEMENT_V1: &'static str = collab_settlement::PROTOCOL;
+
+    const LATEST_PONG: &'static str = Self::PONG_V1;
+    const LATEST_IDENTIFY: &'static str = Self::IDENTIFY_V1;
+    const LATEST_OFFERS: &'static str = Self::OFFERS_V1;
+    const LATEST_ROLLOVER: &'static str = Self::ROLLOVER_V1;
+    const LATEST_COLLAB_SETTLEMENT: &'static str = Self::COLLAB_SETTLEMENT_V1;
```

Comment:
🤔 I don't think these constants give you any guarantees that you are using the correct protocol identifiers in the end. And they kind of obfuscate what's actually going on. I would try to use the actual identifiers directly.

I like the idea of tying the set of listen protocols we say we support with the actual listen protocols we register with the `Endpoint`. It would be better to ask the `Endpoint` about this information, but we have to register the `identify::listener::Actor` too, so there would seem to be a circular dependency there.

I'm gonna try to make sense of what we're doing here. I think this `struct` is trying to do 3 things:

1. Ensuring that taker and maker are compatible at compile-time.
2. Ensuring that the listen protocols we say we support match the ones we run.
3. Allowing us to decide if the maker supports all the listen protocols the taker expects.

I think (1) is somewhat covered by the test `ensure_expected_protocols_are_supported`. We do need to define somewhere what the maker supports and what the taker expects, because they can be different: the maker might support a few more protocols than the taker cares about. But we should be using semantic versioning here, otherwise we will make mistakes in the future.

This makes me think that we should model these protocols a bit like this:

```rust
struct MakerListenProtocols {
    identify: Vec<semver::Version>,
    rollover: Vec<semver::Version>,
    // etc.
}

struct RequiredMakerProtocols {
    identify: semver::VersionReq,
    rollover: semver::VersionReq,
    // etc.
}

fn does_maker_fulfill_taker_needs(maker: MakerListenProtocols, req: RequiredMakerProtocols) -> bool {
    maker.identify.iter().any(|identify| req.identify.matches(&identify)) &&
    maker.rollover.iter().any(|rollover| req.rollover.matches(&rollover)) && 
    // etc.
}
```

We could use this function in the test to ensure compile-time compatibility. And we could also use it for (3) when actually checking compatibility during runtime: `RequiredMakerProtocols` would be set at compile time for the taker and we would have to deserialise the `MakerListenProtocols` during runtime.

Finally (2) could be achieved by adding a method to `MakerListenProtocols` (and a `TakerListenProtocols` that is not specified here but can be imagined) that works similarly to the `{maker,taker}_protocol_handlers` methods in this PR.

In conclusion, I think the intentions and ideas are good, but I would like to see it modelled differently. I'm happy to give it a go myself if you think my proposal is sound.

Reply:
> I don't see the advantage of using semver for the protocol comparison. The taker expects the maker to support the exact same protocol string, otherwise it will not work. As far as I understand protocol negotiation is not semantically versioned but the protocol string has to match exactly - correct me if I'm wrong! That's why I tied this to the actual protocol strings.

I fully expected `multistream-select` to support protocol negotiations using semver but I was wrong! Nothing to action here unfortunately.